algorithms/mopt/mopt_hierarflow_routines.cpp

```diff
@@ -0,0 +1,243 @@
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see http://www.gnu.org/licenses/.
+ */
+
+#include "fuzzuf/algorithms/mopt/mopt_hierarflow_routines.hpp"
+
+#include "fuzzuf/algorithms/mopt/mopt_optimizer.hpp"
+#include "fuzzuf/algorithms/mopt/mopt_option_get_splice_cycles.hpp"
+#include "fuzzuf/utils/common.hpp"
+
+namespace fuzzuf::algorithm::mopt::routine {
+
+namespace other {
+
+using fuzzuf::optimizer::MOptMode;
+using mopt::option::MOptTag;
+
+MOptUpdate::MOptUpdate(MOptState &state) : state(state) {}
+
+MOptMidCalleeRef MOptUpdate::operator()(
+    [[maybe_unused]] std::shared_ptr<MOptTestcase> testcase) {
+  // retrive optimizer values
+  auto last_splice_cycle = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::LastSpliceCycle, true);
+  auto havoc_operator_finds = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::HavocOperatorFinds, true);
+  auto selected_case_histogram = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::SelectedCaseHistogram, true);
+
+  auto &mopt = state.mopt;
+
+  // update havoc_operator_finds
+  for (size_t i = 0; i < havoc_operator_finds.size(); i++) {
+    mopt->accum_havoc_operator_finds[mopt->mode][i] += havoc_operator_finds[i];
+    mopt->accum_selected_case_histogram[mopt->mode][i] +=
+        selected_case_histogram[i];
+  }
+
+  if (last_splice_cycle >= afl::option::GetSpliceCycles(state)) {
+    state.UpdateSpliceCycles();
+  }
+
+  if (mopt->pacemaker_mode) {
+    u64 hit_cnt = state.queued_paths + state.unique_crashes;
+
+    if (unlikely(hit_cnt > hit_cnt * option::GetLimitTimeBound<MOptTag>() +
+                               mopt->pacemaker_hit_cnt)) {
+      mopt->pacemaker_hit_cnt = 0;
+      mopt->pacemaker_mode = false;
+    }
+  }
+
+  auto new_testcases = fuzzuf::optimizer::Store::GetInstance().Get(
+      fuzzuf::optimizer::keys::NewTestcases, true);
+
+  // pilot mode (update local best)
+  if (mopt->mode == MOptMode::PilotMode) {
+    if (unlikely(new_testcases > mopt::option::GetPeriodPilot<MOptTag>())) {
+      auto last_havoc_finds = fuzzuf::optimizer::Store::GetInstance().Get(
+          fuzzuf::optimizer::keys::LastHavocFinds, true);
+      mopt->SetSwarmFitness(last_havoc_finds / new_testcases);
+
+      for (size_t i = 0; i < selected_case_histogram.size(); i++) {
+        double score = 0.0;
+        if (mopt->accum_selected_case_histogram[0][i] > 0) {
+          score = mopt->accum_havoc_operator_finds[0][i] /
+                  mopt->accum_selected_case_histogram[0][i];
+        }
+        mopt->SetScore(i, score);
+      }
+      mopt->UpdateLocalBest();
+      mopt->accum_havoc_operator_finds[0].fill(0);
+      mopt->accum_selected_case_histogram[0].fill(0);
+
+      if (mopt->NextSwarmIdx() == 0) {  // all swarms are visited
+        mopt->UpdateBestSwarmIdx();
+        mopt->mode = MOptMode::CoreMode;
```

Comment:
Sorry for overlooking this, but don't we have to add the code corresponding to this?

https://github.com/puppet-meteor/MOpt-AFL/blob/master/MOpt/afl-fuzz.c#L8732-L8737

Reply:
you're right. fixed in 549e28888267fc779b932b99ac3279199635b3af